Closes #19535 - Add Project Stanza to pyproject toml

pheus opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note that
    our contribution policy requires that a feature request or bug report be
    approved and assigned prior to opening a pull request. This helps avoid
    waste time and effort on a proposed change that we might not be able to
    accept.

    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ISSUE WHICH HAS BEEN ASSIGNED
    TO YOU, IT WILL BE CLOSED AUTOMATICALLY.

    Please specify your assigned issue number on the line below.
-->
### Fixes: #19535 - Update pyproject.toml with a [project] stanza

This PR adds a `[project]` section to `pyproject.toml` to improve compatibility with tools like [`uv`](https://github.com/astral-sh/uv), which rely on PEP 621-compliant metadata.

The section includes basic project metadata such as name, version, description, license, authors, and relevant URLs. This is primarily intended to help users who wish to manage their virtual environments using `uv`, which otherwise fails without a `[project]` table.

The version field is currently maintained manually to match the main NetBox version. Dynamic versioning was considered but isn't suitable here, as we're not building NetBox as a Python package.

Let me know if any adjustments are needed!

pheus committed:
feat(project): Add project metadata to pyproject.toml

Introduces project metadata, including name, version, authors, and
description, to `pyproject.toml` for enhanced package definition.
Also includes URLs for source code, documentation, and issue tracking.

pheus committed:
docs(release): Add checklist item for Python versions in pyproject.toml

Include step to update minimum and supported Python versions
in the project metadata file as part of the release process.

pheus committed:
docs(release): Update checklist to include pyproject.toml versioning

Add a step to update the version in `pyproject.toml` alongside
`release.yaml`.

pheus committed:
feat(project): Update pyproject.toml for best practices

Refreshes metadata to resolve deprecations and follow packaging best
practices. Updates include description, license, Python versions,
classifiers, maintainers, and repository URLs for improved compliance.

jnovinger commented on pyproject.toml:
```diff
@@ -4,38 +4,35 @@
 Homepage = "https://netboxlabs.com/products/netbox/"
-Source = "https://github.com/netbox-community/netbox"
 Documentation = "https://netboxlabs.com/docs/netbox/"
+Repository = "https://github.com/netbox-community/netbox"
```
I saw that the examples use the 'repository' key, but the actual list of [Well-known labels](https://packaging.python.org/en/latest/specifications/well-known-project-urls/#well-known-labels) presents 'source' as the canonical name, with 'repository' as an alias. I think we're fine leaving this one as 'source', unless there's something I'm missing.

pheus replied:
You're absolutely right - `repository` is just an alias for `source`, and `source` is the canonical form per [PEP 621](https://peps.python.org/pep-0621/#source-repository).  
Although technically not best practice, it seems fairly common to use `repository` when referencing a VCS repository. That said, I'm totally fine with keeping it as `source` to align with the canonical format.

jnovinger commented on pyproject.toml:
```diff
@@ -4,38 +4,35 @@
+    "Programming Language :: Python :: 3",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: 3.12",
```
I was looking for best practices for the 'Progamming Language' classifier, but didn't find any right away. Is there a reason for the change to this classifier?

jnovinger reviewed (commented):
This seemed to work pretty well, thanks! I did leave a couple of comments.

For testing, this is the workflow I used, per the [uv docs](https://docs.astral.sh/uv/guides/projects/):

1. Skipped `uv init ...` as pyproject.toml already existed
2. Ran `uv sync` to setup Python version and create .venv/
2. Ran `uv add -r requirements.txt` to add dependencies
4. Tested the `uv run` path with `cd netbox` and `uv run manage.py runserver` to start and test the dev server
5. Tested the 'activate the venv path' with `source .venv/bin/activate`, `cd netbox`, `./manage.py runserver` to start and test the dev server
6. Verified that common workflow steps like running tests and running the `rqworker` management command worked as expected.

Let me know if there's anything else that I should be looking at.

pheus commented:
Wow - thank you so much for the incredibly detailed review and thorough testing notes! 🙌  
Really appreciate the time and care you put into it.

pheus committed:
fix(project): Update repository URL key in pyproject.toml

Replaces the 'Repository' key with 'Source' in accordance with updated
metadata conventions. This ensures compliance with modern best
practices for project metadata.

pheus committed:
fix(project): Specify Python 3 :: Only in classifiers

Updates the Python version classifier in `pyproject.toml` to indicate
support exclusively for Python 3. This change ensures clarity in the
supported Python versions for the project metadata.

jnovinger reviewed (approved):
Thanks, @pheus . I wasn't necessarily opposed to those changes, just wanted to understand them and only include them if they make sense for this work.
